cli/util/uploader.go

```diff
@@ -0,0 +1,434 @@
+	}
+
+	err = updateIndexFiles(ctx, p, containerID, *acc, signer, uint(currentBlockHeight), attr, homomorphicHashingDisabled)
+	if err != nil {
+		return cli.Exit(fmt.Errorf("failed to update index files after upload: %w", err), 1)
+	}
+	return nil
+}
+
+const (
+	maxRetries     = 3                      // Maximum number of retries
+	initialBackoff = 500 * time.Millisecond // Initial backoff duration
+	backoffFactor  = 2                      // Backoff multiplier
+	maxBackoff     = 10 * time.Second       // Maximum backoff duration
+)
+
+// Retry function with exponential backoff.
+func retry(action func() error) error {
+	var err error
+	backoff := initialBackoff
+	for range maxRetries {
+		if err = action(); err == nil {
+			return nil // Success, no retry needed
+		}
+		time.Sleep(backoff) // Backoff before retrying
+		backoff *= time.Duration(backoffFactor)
+		if backoff > maxBackoff {
+			backoff = maxBackoff
+		}
+	}
+	return err // Return the last error after exhausting retries
+}
+
+type searchResult struct {
+	startIndex int
+	numOIDs    int
+	err        error
+}
+
+// fetchLatestMissingBlockIndex searches the container for the last full block batch,
+// starting from the currentHeight and going backwards.
+func fetchLatestMissingBlockIndex(ctx context.Context, p *pool.Pool, containerID cid.ID, priv *keys.PrivateKey, attributeKey string, currentHeight int) (int, error) {
+	var (
+		wg                     sync.WaitGroup
+		maxFullBatchStartIndex = -1
+		numBatches             = currentHeight/searchBatchSize + 1
+	)
+
+	for batch := numBatches; ; batch -= maxParallelSearches {
+		if batch < 0 {
+			batch = 0
+		}
+		results := make([]searchResult, maxParallelSearches)
+
+		for i := range maxParallelSearches {
+			currentBatch := batch - i
+			startIndex := currentBatch * searchBatchSize
+			endIndex := startIndex + searchBatchSize
+
+			if currentBatch <= 0 {
+				startIndex = 0
+				endIndex = searchBatchSize
+			}
+
+			wg.Add(1)
+			go func(i, startIndex, endIndex int) {
+				defer wg.Done()
+
+				prm := client.PrmObjectSearch{}
+				filters := object.NewSearchFilters()
+				filters.AddFilter(attributeKey, fmt.Sprintf("%d", startIndex), object.MatchNumGE)
+				filters.AddFilter(attributeKey, fmt.Sprintf("%d", endIndex), object.MatchNumLT)
+				prm.SetFilters(filters)
+
+				objectIDs, err := neofs.ObjectSearch(ctx, p, priv, containerID.String(), prm)
+				results[i] = searchResult{startIndex: startIndex, numOIDs: len(objectIDs), err: err}
+			}(i, startIndex, endIndex)
+		}
+		wg.Wait()
+
+		for _, res := range results {
+			if res.err != nil {
+				return 0, res.err
+			}
+			if res.numOIDs > 0 {
+				if res.numOIDs == searchBatchSize {
+					maxFullBatchStartIndex = max(maxFullBatchStartIndex, res.startIndex)
+				} else {
+					return res.startIndex, nil
+				}
+			}
+		}
+		if maxFullBatchStartIndex != -1 {
+			return maxFullBatchStartIndex + searchBatchSize, nil
```

Comment:
Be simpler, let's range over results from end to start, and remember that the latest batch may be full. Check this algo:
```suggestion
		for i := len(results) - 1; i >= 0; i-- {
			if results[i].err != nil {
				return 0, results[i].err
			}
			if results[i].numOIDs < searchBatchSize {
				emptyBatchFound = true
				continue
			}
			if emptyBatchFound || (batch == numBatches && i == len(results)-1) {
				// Return index of the first block in the first half-empty bath.
				return results[i].startIndex + searchBatchSize, nil
			}
		}
```